design-proposals/default-instancetypes-from-volumes.md

```diff
@@ -0,0 +1,86 @@
+
+## Goals
+
+* Allow users to opt into having a default instancetype and preference inferred from a given `Volume` associated with their `VirtualMachine`.
+
+## Non Goals
+
+* Automatic annotation of a `Volume` or underlying type (such as a `PVC`) is not covered as part of this work but is something that will be investigated in the future.
+
+## User Stories
+
+* As an admin, user or third party creator of a `Volume` (for example the `CDI` project using `DataVolumes` and `PVCs`),  I want to provide annotations that suggest the default instancetype or set of preferences.
+
+* As a user, I want to opt into allowing KubeVirt to infer these defaults from a volume associated with my VirtualMachine.
+
+## Repos
+
+* [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+
+The `Volume` type will be extended to include a new `inferInstancetype` and `inferPreference` boolean attributes that will control this behaviour, defaulting to `False`.
+
+The attribute will only be allowed against a single `Volume` associated with a `VirtualMachine`, with this behaviour enforced early within the API by the mutation webhook before any additional logic is executed.
+
+When enabled against a single `Volume` KubeVirt will then look at the underlying `Volume` type (such as a `PVC`) for the following annotations:
+
+* `instancetype.kubevirt.io/defaultInstancetype`
+* `instancetype.kubevirt.io/defaultInstancetypeKind` (Defaults to `VirtualMachineClusterInstancetype`)
+* `instancetype.kubevirt.io/defaultPreference`
+* `instancetype.kubevirt.io/defaultPreferenceKind` (Defaults to `VirtualMachineClusterPreference`)
+
+If found these annotations will be used to populate the `InstancetypeMatcher` and `PreferenceMatcher` of the `VirtualMachine` by the mutation webhook with the existing checks of the validation webhook then used to ensure these resources exist and apply without conflicts to the `VirtualMachine` and eventual `VirtualMachineInstance` at runtime.
+
+## API Examples
+
+```yaml
+$ wget https://github.com/cirros-dev/cirros/releases/download/0.5.2/cirros-0.5.2-x86_64-disk.img
+[..]
+$ ./cluster-up/virtctl.sh image-upload pvc cirros --size=1Gi --image-path=./cirros-0.5.2-x86_64-disk.img
+[..]
+$ ./cluster-up/kubectl.sh kustomize https://github.com/lyarwood/common-instancetypes.git | ./cluster-up/kubectl.sh apply -f -
+[..]
+$ ./cluster-up/kubectl.sh annotate pvc/cirros instancetype.kubevirt.io/defaultInstancetype=server.medium instancetype.kubevirt.io/defaultPreference=rhel.7
+[..]
+$ cat <<EOF | ./cluster-up/kubectl.sh apply -f -
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: cirros
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        devices: {}
+      volumes:
+      - persistentVolumeClaim:
+          claimName: cirros
+        name: disk
+        inferInstancetype: true
+        inferPreference: true
```

Comment:
If we say that IT and Pref will be infered from one volume only, then we could do the following instead:

```
kind: VM
spec:
  inferFromVolumeName: disk
```

With this IT/Pref does not "leak" to the VMI.
It would also change the semantics to make it implicit that only one volume can be selected for the inference.

Reply:
Yeah that's valid, or we can extend the `{Instancetype,Preference}Matcher` so it's defined in the normal place within the `VirtualMachineSpec`. Something like:

```yaml
kind: VirtualMachine
spec:
  instancetype:
    inferFromVolume: disk
  preference:
    inferFromVolume: disk
```

Reply:
Yep, that loosk nice.

Btw another approach would be to have a boolean on the volume (not disk).

Or, have a convention: The first volume is the source for inst/pref, can optionally be overriden by a mechanism as descibred by you in the comment above.

Reply:
I'd prefer to have it in the matcher. A default behaviour of picking the first volume is also valid but it will add at least one additional lookup of the first volume to each VirtualMachine creation flow by default. I wonder if it's worth it?